lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1122,6 +1122,11 @@ static bool performCompile(CompilerInstance &Instance,
   if (Action == FrontendOptions::ActionType::Typecheck) {
     if (emitIndexData(Invocation, Instance))
       return true;
+    // FIXME: Whole-module outputs with a non-whole-module -typecheck ought to
+    // be disallowed, but indexing still relies on taking a regular command line
+    // and making minimal modifications to it to index just one file instead.
+    // For now, guard the emission of whole-module supplementary outputs with a
+    // check that we are actually in a whole-module mode.
```

Comment:
Thanks for doing this! To my eye, the comment above could be even clearer. Here's the sort of thing I have in mind. Feel free to take it, modify it, or leave it:

Xcode [or is it the driver?] generates an index [-`emit-index`?] by taking a regular, command line with `-wmo` and modifies it to index just one file instead of every source file. If that line also has flags causing the compiler to emit whole-module supplementary outputs, the compiler can crash trying to access information for non-type-checked declarations in the non-primary files. Prevent those crashes by guarding the emission of whole-module supplementary outputs.